neuraxle/metaopt/random.py

```diff
@@ -114,6 +117,103 @@ def _split(self, data_inputs):
         return splitted_data_inputs
 
 
+class WalkForwardTimeSeriesCrossValidation(BaseCrossValidation):
+    """
+    This will change perform a walk forward cross validation by performing a forward rolling split.
+    All the training split will start at the start of the time series but the end will increase toward
+    the end at each forward step. For the validation split it will start after each training split and
+    can have a delay to test prediction according to some delay.
+
+    :param scoring_function: scoring function use to validate performance if it is not None, by default r2_score,
+    :param window_size: the size of the time step taken between each forward roll, by default 3.
+    :param initial_window_number: the size of the first training split length in term of number of window a window is defined by the window_size, by default 3.
+    :param window_delay_size: the size of the delay between the end of the training split and the validation split, by default 0.
+    :param joiner the joiner callable that can join the different result together.
+    :return: WalkForwardTimeSeriesCrossValidation instance.
+    """
+
+    def __init__(self, scoring_function=r2_score, window_size=3, initial_window_number=3, window_delay_size=0, joiner=NumpyConcatenateOnTimeStep()):
+        self.window_size = window_size
+        self.window_delay_size = window_delay_size
+        self.initial_window_number = initial_window_number
+        super().__init__(scoring_function=scoring_function, joiner=joiner)
+
+    def split(self, data_inputs, expected_outputs):
+        """
+        Split the data into train inputs, train expected outputs, validation inputs, validation expected outputs
+        :param data_inputs: data to perform walk forward cross validation into.
+        :param expected_outputs: the expected target*label that will be used during walk forward cross validation.
+        :return: train_data_inputs, train_expected_outputs, validation_data_inputs, validation_expected_outputs
+        """
+        # TODO: Verify that we need to split into training and validation or into inputs and expected_outputs.
```

Comment:
I think that in an autoregressive autoencoder: 

- "training" split is similar to "data_inputs"
- "validation" split is similar to "expected_outputs"

That is the case at the very least, for which the expected output is somehow the validation set. It's just that the split will re-happen at a level lower (so that the training split will be itself splitted into a bunch of data_inputs and expected outputs, and that the test split will also be itself splitted into a bunch of data_inputs and expected outputs too).